### DISCUSSION TITLE
Supporting SentencePiece tokenization algorithms other than BPE

### DISCUSSION THREAD

**Discussion by fairydreaming** [ggml-org/llama.cpp#7732] [Ideas]
I'd like to discuss what changes are needed in llama.cpp to support SentencePiece tokenization algorithms other than BPE. 
Let me start with a short introduction to the problem:

1. There are four tokenization algorithms available in SentencePiece library: BPE, unigram, char, word.
2. In llama.cpp there is a [llm_tokenizer_spm](https://github.com/ggerganov/llama.cpp/blob/bde7cd3cd949c1a85d3a199498ac98e78039d46f/llama.cpp#L12726-L12850) tokenizer that is used for LLAMA_VOCAB_TYPE_SPM. While its name sounds like a kind of "generic" sentencepiece tokenizer, from my understanding it implements only the BPE tokenization algorithm.
3. Information about the tokenization algorithm used to create SentencePiece model is currently not extracted by convert-hf, so it's not stored in a model header and llama.cpp has no way to know about the used SentencePiece tokenization algorithm.
4. This will cause the llama.cpp to silently use the incorrect tokenization algorithm if a non-BPE tokenization algorithm (unigram, char, word) was used to create the SentencePiece tokenizer model.

I know that non-BPE tokenization algorithms are rarely used, but for example T5 model uses SentencePiece unigram tokenization algorithm and adding T5 support is on [llama.cpp roadmap](https://github.com/users/ggerganov/projects/7/views/1). There was also at least one bug report caused by this issue: #6717.

I think the following steps are needed to fix this problem:

1. Add reading information about used tokenization algorithm in convert-hf. This information seems to be stored in [TrainerSpec model_type field](https://github.com/google/sentencepiece/blob/b9ecf1ceed910be5ccf48d4602824d44ba59397b/src/sentencepiece_model.proto#L54) in SentencePiece tokenizer model file.
2. Add information about detected tokenization algorithm to the model header. There is a model header field tokenizer.ggml.model that is currently used to store the information about the tokenizer type. I noticed that the set of currently used values of this field is somewhat unfortunate ("llama", "bert", "gpt2"). Is there any particular reason why we have model names here instead of actual tokenization algorithm names like "bpe", "wordpiece" etc?
3. Add new values in [llama_vocab_type enum](https://github.com/ggerganov/llama.cpp/blob/3b38d48609280aa5f8ab7ea135a4351b2a5ee240/llama.h#L65-L70) for four SentencePiece tokenization algorithms (BPE, unigram, char, word) instead of current single LLAMA_VOCAB_TYPE_SPM.
4. Add support for SentencePiece unigram, char and word tokenization algorithms in llama.cpp. Initially it could simply fail with information that given tokenization algorithm is not yet supported.

Let me know what do you think about this.

**Comment by ggerganov** [comment]
> I noticed that the set of currently used values of this field is somewhat unfortunate ("llama", "bert", "gpt2"). Is there any particular reason why we have model names here instead of actual tokenization algorithm names like "bpe", "wordpiece" etc?

I guess we didn't fully understand the tokenization variants back when we designed this spec:

https://github.com/ggerganov/ggml/pull/302#issuecomment-1639893618

For a long time, I thought that SPM and BPE are two different tokenization algorithms.

We should be able to extend `tokenizer.ggml.model` and pass proper tokenizer type strings (such as "bpe", "unigram", "char", etc.) - we just need to handle the existing strings as well for backwards compatibility.

Would be great to support more tokenizer types, especially for various embedding and T5 models that we don't currently support. Contributions in that direction are welcome

**Reply by fairydreaming** [reply]
@ggerganov While LLAMA_VOCAB_TYPE_SPM and LLAMA_VOCAB_TYPE_BPE both use BPE tokenization algorithm, there are differences between the two implementations (for example usage of scores in SPM vs ranks in BPE, pre-tokenization with regex in BPE), I think we should probably keep them separated. This would suggest that not only the type of tokenization algorithm is important, but also its specific implementation. As for the Unigram tokenization, I found the following in [this paper](https://arxiv.org/html/2310.08754v4):

> Our studies rely on the two established tokenization algorithms, BPE and Unigram, and their implementation in the Huggingface tokenizer library Moi and Patry ([2023](https://arxiv.org/html/2310.08754v4#bib.bib35)) and the SentencePiece library Kudo and Richardson ([2018](https://arxiv.org/html/2310.08754v4#bib.bib30)). We considered both libraries in order to investigate the effect of differences in the pre-and post-processing steps and potential differences in the implementations. Due to missing pre-processing options for Huggingface’s Unigram implementation, which causes a large discrepancy in the resulting vocabulary compared to SentencePiece’s implementation of Unigram, we omitted the training of Unigram tokenizers based on Huggingface.

This would suggest that Unigram implementations in HF tokenizer library and SentencePiece are also slightly different. So I guess the following information shall be present in tokenizer.ggml.model:

1. Tokenization algorithm type (wordpiece, bpe, unigram etc)
2. Used implementation of the tokenization algorithm (SentencePiece vs HF tokenizers library)

So perhaps values of tokenizer.ggml.model field shall be something like: bpe-spm, bpe-hft, unigram-spm, unigram-hft, wordpiece-hft etc. Alternatively there could be another field tokenizer.ggml.implementation for keeping the information about used tokenizer implementation.

I am by no means a tokenization expert, so I think someone more knowledgeable should confirm this.
@teleprint-me it looks like you are currently working on a PR that is related to this, what do you think?

**Reply by ggerganov** [reply]
> While LLAMA_VOCAB_TYPE_SPM and LLAMA_VOCAB_TYPE_BPE both use BPE tokenization algorithm, there are differences between the two implementations (for example usage of scores in SPM vs ranks in BPE, pre-tokenization with regex in BPE), I think we should probably keep them separated. 

Yes, got it

> This would suggest that not only the type of tokenization algorithm is important, but also its specific implementation.

Ugh, not sure. It can get too complicated if we start supporting different implementations. Probably stick to the HF tokenizer as it is the most widely used one?

**Reply by ggerganov** [reply]
cc @jaime-m-p for any additional insights on this topic

**Reply by fairydreaming** [reply]
> > While LLAMA_VOCAB_TYPE_SPM and LLAMA_VOCAB_TYPE_BPE both use BPE tokenization algorithm, there are differences between the two implementations (for example usage of scores in SPM vs ranks in BPE, pre-tokenization with regex in BPE), I think we should probably keep them separated.
> 
> Yes, got it
> 
> > This would suggest that not only the type of tokenization algorithm is important, but also its specific implementation.
> 
> Ugh, not sure. It can get too complicated if we start supporting different implementations. Probably stick to the HF tokenizer as it is the most widely used one?

@ggerganov But you already do support different implementations, as is the case with LLAMA_VOCAB_TYPE_SPM (SentencePiece BPE) vs LLAMA_VOCAB_TYPE_BPE (HuggingFace tokenizers BPE). I don't think we have any control over this, it's a result of the choice made by LLM model creators.

**Reply by ggerganov** [reply]
> But you already do support different implementations, as is the case with LLAMA_VOCAB_TYPE_SPM (SentencePiece BPE) vs LLAMA_VOCAB_TYPE_BPE (HuggingFace tokenizers BPE). I don't think we have any control over this, it's a result of the choice made by LLM model creators.

> So perhaps values of tokenizer.ggml.model field shall be something like: bpe-spm, bpe-hft, unigram-spm, unigram-hft, wordpiece-hft etc.

Oh, I see - I think this makes sense

**Reply by teleprint-me** [reply]
I'm no expert. I'm literally learning as I go. GPT-4 taught me the basics and I went from there. I deliver food for money, 😅. I'm terrible at math (even though I find it fascinating), but I have been programming since I was 15. Now that I've relieved my self.

I'm in the middle of doing what you're talking about here. I'm building a bridge between HF and llama.cpp. The `class Tokenizer` is a big part of that as it's the first step in most cases for most of us users.

Things that are missing for the tokenizer are:

- knowing what architecture it belongs to, llama, falcon, gpt2, etc. I used `tokenizer.model` for this.
- knowing what tokenizer type was used, bpe, wpm, unigram, ngram, etc. I used `tokenizer.type` for this.
- per token metadata, e.g. whether to use spaces, prefix, postfix, etc. This information can be partitioned into the model file. e.g. `tokenizer.pre`, `tokenizer.vocab`, etc.

A lot of this is just metadata provided by `tokenizer.json`.

Sentencepiece is more involved because google did google things. They used the proto interface which is currently alien to me, but if I had the money and time, I could dig into it.

I'm focused mostly on BPE because it's the most widely used in the majority of cases from what I've found. Hugging Face is streamlining their model interface, so I expect it to converge over time. Their SPM support is terrible and the conversion for SPM is broken.

Implementing a BPE tokenizer in C for llama.cpp should actually be pretty easy in isolation. It's the integration support that's challenging, but not impossible; It's definitely doable.

Without a tokenizer, there is no LLM, no inference, no training. It's the heart of LLMs. Understanding the tokenizer is crucial. It pumps the numerical values through the LLM making prompt processing and token generation possible. Otherwise, there is nothing to graph.

**Reply by fairydreaming** [reply]
@teleprint-me 

> I'm in the middle of doing what you're talking about here. I'm building a bridge between HF and llama.cpp. The `class Tokenizer` is a big part of that as it's the first step in most cases for most of us users.
> 
> Things that are missing for the tokenizer are:
> 
>     * knowing what architecture it belongs to, llama, falcon, gpt2, etc. I used `tokenizer.model` for this.

I'm not sure why is it needed. Is it any different from general.architecture field?

>     * knowing what tokenizer type was used, bpe, wpm, unigram, ngram, etc. I used `tokenizer.type` for this.

Are going to use a combination of model and type values to determine what llama.cpp tokenizer class shall be used?

**Reply by teleprint-me** [reply]
Honestly, I haven't decided yet as I'm still working my way there. As time passes and more users become reliant upon llama.cpp, breakage becomes more damaging.

The current implementation in my PR and experimental repo is simply to play around with different potential implementations before committing to anything. I'm open to alternative approaches.

> I'm not sure why is it needed. Is it any different from general.architecture field?

No, it's not any different than the general architecture field as it's simply to know which tokenizer is associated with which model architecture. The idea here was to enable future compatibility for training tokenizers in isolation. This would allow users to create custom tokenizers with llama.cpp and then later train a language model in llama.cpp with that tokenizer.

> Are going to use a combination of model and type values to determine what llama.cpp tokenizer class shall be used?

I had considered using a combination of model and type values, yes. I thought it might make it easier to identify the tokenizer class to use, but that's yet to be proven. I won't feel confident in this approach until I have it more fleshed out. I have less confidence in theory than I do in application.

**Reply by fairydreaming** [reply]
> > Are going to use a combination of model and type values to determine what llama.cpp tokenizer class shall be used?
> 
> I had considered using a combination of model and type values, yes. I thought it might make it easier to identify the tokenizer class to use, but that's yet to be proven. I won't feel confident in this approach until I have it more fleshed out. I have less confidence in theory than I do in application.

@teleprint-me I see some problems with this approach, for example currently:
- "llama" tokenizer.ggml.model value used in LLaMA and LLaMA-2 maps to LLAMA_VOCAB_TYPE_SPM which maps to llm_tokenizer_spm (SentencePiece BPE tokenizer)
- "gpt2" tokenizer.ggml.model value used in LLaMA-3 maps to LLAMA_VOCAB_TYPE_BPE which maps to llm_tokenizer_bpe (HuggingFace tokenizers BPE tokenizer)

In your approach tokenizer.model value in both these cases will be "llama" and tokenizer.type in both cases will be "bpe". So you will have to use other means to distinguish between the two.

**Reply by teleprint-me** [reply]
> In your approach tokenizer.model value in both these cases will be "llama" and tokenizer.type in both cases will be "bpe". So you will have to use other means to distinguish between the two.

Yes, I'm aware of these issues. That's what I'm currently working on figuring out. I see an opportunity for improvement, clarity, and a potentially robust implementation.

@jaime-m-p has already laid out some potential ground work towards this end.

**Reply by jaime-m-p** [reply]
> It can get too complicated if we start supporting different implementations. Probably stick to the HF tokenizer as it is the most widely used one?

I'm not an expert in tokenizers neither and I don't know the differences between implementation variants, but having the base algorithms and variables/flags probably we only need to add some *if*s to handle the differences.

I agree to start miming HF tokenizers and then add the variants once we have a solid tested base.

**Reply by teleprint-me** [reply]
My current thoughts are:

- The unicode and regex are really the most difficult components; these are in place already. 
- The front-end API seems to have generic interfaces at the moment; This is in progress. 
- I think the BPE implementation is partial and it's the easiest to implement. [Paper here](https://arxiv.org/abs/1508.07909v5). 

Some stemming thoughts are:

- This tackles two tasks at once; BPE and HF.
- Word Piece does differ and this will take some time.
- We can use HF as a control to see if we're on the right track.

The implementation should be straightforward once the metadata is set. A possible approach for the metadata is to partition it in a two-fold approach.

- Consider what information is needed for the current SPM implementation.
- Consider what information is needed for the in progress HF implementation.

Then partition the `Tokenizer` metadata accordingly with a consistent interface. We can follow the HF setup and use a mapping in that regard. The tools are there already, e.g. `common/json.hpp`.

**Comment by teleprint-me** [comment]
Mentioning Iss #7763 here as it's connected. @ggerganov @slaren and @fairydreaming might find the matmul free paper I mentioned there interesting. The algorithm is included in the paper. Almost forgot, the quant algorithm is in the appendix.

tagging @jart and @JohannesGaessler for matmul free to get some extra eyes on it. I have no idea how valid the math is. Wouldn't mind some feedback on it.

I think this is all interconnected simply because having a vocab free and matmul free llm should be obviously advantageous.
